### PR TITLE
IBC fixes

### PR DESCRIPTION
**Pull request by davidwrighton**
- Support IntPtr and UIntPtr element types in IBC data
- More defensive coding against structurally incorrect IBC data
- Skip compilation of open method definitions
- Generate IBC data without crashing when the first logged item is an InstantiatedMethod and not a token
- Safely shutdown and emit IBC data from a thread that hasn't been given a Thread object

### CODE REVIEW

File: src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunLibraryRootProvider.cs [hunk 1]
```diff
@@ -37,6 +37,12 @@ public void AddCompilationRoots(IRootingServiceProvider rootProvider)
                         {
                             continue;
                         }
+
+                        if (method.IsMethodDefinition)
```
Comment (briansull) [inline src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunLibraryRootProvider.cs]: Add a comment explaining why you added this.

File: src/vm/ceemain.cpp [hunk 2]
```diff
@@ -1449,8 +1449,18 @@ void STDMETHODCALLTYPE EEShutDownHelper(BOOL fIsDllUnloading)
...
                 if (g_IBCLogger.InstrEnabled())
                 {
-                    Thread * pThread = GetThread();
-                    ThreadLocalIBCInfo* pInfo = pThread->GetIBCInfo();
+                    Thread * pThread = GetThreadNULLOk();
+                    ThreadLocalIBCInfo* pInfo = NULL;
+
+                    if (pThread != NULL)
+                    {
+                        if (pInfo == NULL)
```
Comment (briansull) [inline src/vm/ceemain.cpp]: Yes it will be NULL

File: src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunLibraryRootProvider.cs (hunk 1 above)
Comment (davidwrighton) [reply src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunLibraryRootProvider.cs]: The comment actually already exists above. This is about not working with uninstantiated methods.

File: src/vm/ceemain.cpp (hunk 2 above)
Comment (davidwrighton) [reply src/vm/ceemain.cpp]: Oops, it shouldn't be. I missed a line when copying this change from my work branch.

**Pull request merged**